### PR TITLE
only copy the model once when predicting multiple batches

### PR DESCRIPTION
**Pull request by rongou**
@canonizer @RAMitchell @sriramch

### CODE REVIEW

File: src/predictor/gpu_predictor.cu [hunk 1]
```diff
@@ -268,28 +268,30 @@ class GPUPredictor : public xgboost::Predictor {
...
+        tree_segments_.resize(h_tree_segments.size());
+        dh::safe_cuda(cudaMemcpyAsync(dh::Raw(tree_segments_), h_tree_segments.data(),
+                                      sizeof(size_t) * h_tree_segments.size(),
+                                      cudaMemcpyHostToDevice));
+
+        tree_group_.resize(model.tree_info.size());
+        dh::safe_cuda(cudaMemcpyAsync(dh::Raw(tree_group_), model.tree_info.data(),
+                                      sizeof(int) * model.tree_info.size(),
+                                      cudaMemcpyHostToDevice));
+      }
```
Comment (canonizer) [inline src/predictor/gpu_predictor.cu]: This code would look better as a separate method. In my opinion, this looks more logical, and reduces the number of function parameters.

Comment (rongou) [reply src/predictor/gpu_predictor.cu]: Done.

File: src/predictor/gpu_predictor.cu [hunk 2]
```diff
@@ -361,10 +364,14 @@ class GPUPredictor : public xgboost::Predictor {
       DeviceOffsets(batch.offset, batch.data.Size(), &device_offsets);
       batch.data.Reshard(GPUDistribution::Explicit(devices_, device_offsets));
 
-      // TODO(rongou): only copy the model once for all the batches.
+      if (batch_offset == 0) {
+        dh::ExecuteIndexShards(&shards_, [&](int idx, DeviceShard& shard) {
+          shard.InitModel(model, h_tree_segments, h_nodes);
+        });
+      }
```
Comment (canonizer) [inline src/predictor/gpu_predictor.cu]: Consider moving it outside of the loop. You won't need the conditional then.

File: src/predictor/gpu_predictor.cu [hunk 3]
```diff
@@ -262,35 +262,38 @@ class GPUPredictor : public xgboost::Predictor {
...
-
       dh::safe_cuda(cudaMemcpyAsync(dh::Raw(tree_group_), model.tree_info.data(),
                                     sizeof(int) * model.tree_info.size(),
                                     cudaMemcpyHostToDevice));
+    }
 
+    void PredictInternal
+    (const SparsePage& batch, const MetaInfo& info,
+     HostDeviceVector<bst_float>* predictions,
+     size_t tree_begin, size_t tree_end, int n_classes) {
```
Comment (canonizer) [inline src/predictor/gpu_predictor.cu]: Why do you need `n_classes`?

File: src/predictor/gpu_predictor.cu [hunk 4]
```diff
@@ -361,10 +364,14 @@ class GPUPredictor : public xgboost::Predictor {
       DeviceOffsets(batch.offset, batch.data.Size(), &device_offsets);
       batch.data.Reshard(GPUDistribution::Explicit(devices_, device_offsets));
 
-      // TODO(rongou): only copy the model once for all the batches.
+      if (batch_offset == 0) {
```
Comment (sriramch) [inline src/predictor/gpu_predictor.cu]: can this be hoisted out of the for loop?

File: src/predictor/gpu_predictor.cu [hunk 5]
```diff
@@ -361,10 +364,14 @@ class GPUPredictor : public xgboost::Predictor {
...
+        dh::ExecuteIndexShards(&shards_, [&](int idx, DeviceShard& shard) {
```
Comment (canonizer) [inline src/predictor/gpu_predictor.cu]: Given that the code processing the tree nodes above (lines 335-350) has to do with model initialization, consider moving it (together with calls to `DeviceShard::InitModel()`) to a separate method of GPUPredictor.

File: src/predictor/gpu_predictor.cu (hunk 3 above)
Comment (rongou) [reply src/predictor/gpu_predictor.cu]: It's needed by the prediction kernel.

File: src/predictor/gpu_predictor.cu (hunk 4 above)
Comment (rongou) [reply src/predictor/gpu_predictor.cu]: Done.

File: src/predictor/gpu_predictor.cu (hunk 2 above)
Comment (rongou) [reply src/predictor/gpu_predictor.cu]: Done.

File: src/predictor/gpu_predictor.cu (hunk 5 above)
Comment (rongou) [reply src/predictor/gpu_predictor.cu]: Done.

File: src/predictor/gpu_predictor.cu [hunk 6]
```diff
@@ -361,10 +371,9 @@ class GPUPredictor : public xgboost::Predictor {
       DeviceOffsets(batch.offset, batch.data.Size(), &device_offsets);
       batch.data.Reshard(GPUDistribution::Explicit(devices_, device_offsets));
 
-      // TODO(rongou): only copy the model once for all the batches.
       dh::ExecuteIndexShards(&shards_, [&](int idx, DeviceShard& shard) {
-        shard.PredictInternal(batch, dmat->Info(), out_preds, model,
-                              h_tree_segments, h_nodes, tree_begin, tree_end);
+        shard.PredictInternal(batch, dmat->Info(), out_preds, tree_begin, tree_end,
+                              model.param.num_output_group);
```
Comment (canonizer) [inline src/predictor/gpu_predictor.cu]: I think all three of these parameters can be stored in the shard after `InitModel()`.

However, I'll leave this up to you.

File: src/predictor/gpu_predictor.cu (hunk 3 above)
Comment (rongou) [reply src/predictor/gpu_predictor.cu]: Done.

File: src/predictor/gpu_predictor.cu [hunk 7]
```diff
@@ -321,13 +324,7 @@ class GPUPredictor : public xgboost::Predictor {
...
   };
 
-  void DevicePredictInternal(DMatrix* dmat,
-                             HostDeviceVector<bst_float>* out_preds,
-                             const gbm::GBTreeModel& model, size_t tree_begin,
-                             size_t tree_end) {
-    if (tree_end - tree_begin == 0) { return; }
-    monitor_.StartCuda("DevicePredictInternal");
-
+  void InitModel(const gbm::GBTreeModel &model, size_t tree_begin, size_t tree_end) {
```
[the remark this answers is not in the document]
Comment (rongou) [reply src/predictor/gpu_predictor.cu]: Done.

File: src/predictor/gpu_predictor.cu (hunk 6 above)
Comment (rongou) [reply src/predictor/gpu_predictor.cu]: Done.

**Comment by rongou** [comment]
@RAMitchell this is ready to merge. Thanks!

**Comment by hcho3** [comment]
@RAMitchell  Do we want this in 0.90?

**Comment by hcho3** [comment]
@rongou @canonizer @sriramch Can you do me a favor and explain what this PR does? Is this a follow-up to #4284 (external memory with single GPU) and #4438 (external memory with multiple GPUs)?

**Comment by rongou** [comment]
@hcho3 yes it's an optimization and some refactoring. In external memory mode when we are running prediction on multiple batches, we should only copy the model to GPU once instead of every batch.

**Comment by RAMitchell** [comment]
@hcho3 this is a fairly low impact change, up to you whether you want to include it. I will merge so as not to have prs sitting around.

**Pull request merged**